src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioErrorReportingService.cs

```diff
@@ -0,0 +1,113 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Extensions;
+using Microsoft.VisualStudio.Imaging;
+using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using VS = Microsoft.VisualStudio.Progression;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation
+{
+    internal class VisualStudioErrorReportingService : IErrorReportingService
+    {
+        private VisualStudioWorkspaceImpl workspace;
+        private static InfoBarButton EnableItem = new InfoBarButton(ServicesVSResources.Enable);
+        private static InfoBarButton EnableAndIgnoreItem = new InfoBarButton(ServicesVSResources.EnableAndIgnore);
+
+        public VisualStudioErrorReportingService(VisualStudioWorkspaceImpl workspace)
+        {
+            this.workspace = workspace;
+        }
+
+        public async void ShowErrorInfoForCodeFix(DocumentId documentId, string codefixName, Action OnEnableClicked, Action OnEnableAndIgnoreClicked)
+        {
+            // We can be called from any thread since errors can occur anywhere, however we can only construct and InfoBar from the UI thread.
+            // We try and get the current UI dispatcher and invoke our code on that thread. We return immediately so we don't block waiting on the UI thread.
+            var dispatcher = VS.UIThread.GetUIDispatcher();
+            await dispatcher?.InvokeAsync(() =>
+            {
+                IVsWindowFrame frame;
+                IVsInfoBarUIFactory factory;
+                if (workspace.TryGetInfoBarData(documentId, out frame, out factory))
+                {
+                    CreateInfoBar(codefixName, OnEnableClicked, OnEnableAndIgnoreClicked, frame, factory);
```

Comment:
what is life time of the lambda? what happen if the lambda has captured variables?